tainting: Add experimental taint labels feature

IagoAbal opened the pull request:
It provides the ability to attach labels to different kinds of taint.
Both sources and sinks can retrict what labels are present in the data
that passes through them in order to apply. This allows to write more
complex taint rules that previously required ugly workarounds. It also
allows us to write certain classes of typestate analyses.

Closes PA-1362

test plan:
make test # added 7 tests

PR checklist:

- [x] Documentation is up-to-date
- [x] `changelog.d/<issue>.<type>` is a file with the _what_, _why_, and _how_ of the change.
  - \<issue> is `pa-312` (Linear ticket), `gh-1234` (GitHub issue), or `new-gizmo` (unique semantic name)
  - \<type> is `added`, `changed`, `fixed`, or `infra`.
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see the [Contribution guidelines](https://semgrep.dev/docs/contributing/how-to-contribute/)!

aryx commented on changelog.d/pa-1362.added:
```diff
@@ -0,0 +1,5 @@
+Add experimental support for _taint labels_, that is the ability to attach labels to
+different kinds of taint. This allows to write more complex taint rules that
+previously required ugly workarounds. Taint labels also allows us to write certain
```
typo: us -> useful

IagoAbal commented:
Documentation ticket: [PA-1686](https://linear.app/r2c/issue/PA-1686)

aryx commented on semgrep-core/src/core/Rule.ml:
```diff
@@ -167,15 +167,30 @@ type pformula = New of formula | Old of formula_old [@@deriving show, eq]
+  formula : pformula;
+  label : string;  (** The label to attach to the data *)
+  requires : AST_generic.expr;
+      (** A Boolean formula over taint labels, the expression that
```
please do not use "formula" as we already use it for the formula/combination of patterns.

aryx commented on semgrep-core/src/core/Rule.ml:
```diff
@@ -167,15 +167,30 @@ type pformula = New of formula | Old of formula_old [@@deriving show, eq]
+      (** A Boolean formula over taint labels, the expression that
+       is being checked as a source must satisfy this in order to the
+       label to be produced. Note that with 'requires' a taint source
+       behaves a bit like a propagator ... *)
```
Maybe you can add here that the expressions will be evaluated at matching time by Eval_generic?

IagoAbal committed:
tainting: Add experimental taint labels feature

It provides the ability to attach labels to different kinds of taint.
Both sources and sinks can retrict what labels are present in the data
that passes through them in order to apply. This allows to write more
complex taint rules that previously required ugly workarounds. It also
allows us to write certain classes of typestate analyses.

Closes PA-1362

test plan:
make test # added 7 tests

aryx commented on semgrep-core/src/core/Rule.ml:
```diff
@@ -167,15 +167,30 @@ type pformula = New of formula | Old of formula_old [@@deriving show, eq]
-type sanitizer_spec = {
+type taint_source = {
+  formula : pformula;
+  label : string;  (** The label to attach to the data *)
```
Should it be an option? Or do you have a default value for it?
In that case maybe the default value should be defined in this file.

aryx commented on semgrep-core/src/parsing/Parse_rule.ml:
```diff
@@ -810,6 +812,8 @@ let parse_extract_reduction ~id (s, t) =
 (* Sub parsers taint *)
 (*****************************************************************************)
 
+let builtin_label_SOURCE = "__SOURCE__"
```
I would move that to Rule.ml

aryx commented on semgrep-core/src/tainting/Dataflow_tainting.ml:
```diff
@@ -160,23 +171,71 @@ let merge_source_sink_mvars env source_mvars sink_mvars =
+         | Arg _ -> None)
+  |> LabelSet.of_seq
+
+let rec eval_label_requires ~labels e =
```
maybe add a (* coupling: if you modify the code here, you probably need to modify Parse_rule.parse_require_expression *)

aryx commented on semgrep-core/tests/OTHER/rules/taint_labels2.yaml:
```diff
@@ -0,0 +1,17 @@
+    message: |
+      This confirms taint mode works.
+    pattern-sinks:
+      - requires: TAINTED and not CLEANED
```
Does that mean sanitizer are now useless? Or they are still useful as a quicker way to express
source/sanitize/sink instead of having to invent labels and requires expressions?

IagoAbal replied:
I am not sure to be honest, I think we'll have to gain some experience in using labels. Sanitizers do allow you to _completely remove_ taint from data, although I think that can be simulated with labels too. I could have translated (internally) `pattern-sanitizers` to `patter-sources` with a `"__SANITIZED__"` label, and then rewrite the `requires` of sinks as `(...) not __SANITIZED__`. But `not-conflicting` sanitizers cannot be simulated that way. Although I think we should replace `not-conflicting` with some `options:` because they are a bit confusing to use sometimes.

aryx commented on semgrep-core/tests/OTHER/rules/taint_labels5.yaml:
```diff
@@ -0,0 +1,23 @@
+              $FILE = open(...)
+          - focus-metavariable: $FILE
+    pattern-sinks:
+      - requires: closed and not reopened
```
noice!

IagoAbal committed:
Yoann's review
